File: Codes/Hash/Non-Cryptographic/ELFHash/ELFHash.c
/*
    Modifikasi PJWHash untuk sistem UNIX.
    Archive of Reversing.ID
    Non-Cryptographic Hash

    Assemble:
        (gcc)
        $ gcc -m32 -S -masm=intel -o ELFHash.asm ELFHash.c

        (msvc)
        $ cl /c /FaELFHash.asm ELFHash.c
*/
#include <stdint.h>

uint32_t ELFHash (const char* key, uint32_t length)
{
    uint32_t test = 0;
    uint32_t state = 0;
    uint32_t i = 0;

    for (i = 0; i < length; ++key, i++)
    {
        state = (state << 4) + (*key);

        test = state & 0xF0000000;
        if (test)
        {
            state ^= (test >> 24);
        }
        state &= ~test;
    }
    return state;
}

File: Codes/Hash/Non-Cryptographic/ELFHash/ELFHash.java
/*
	Modifikasi PJWHash untuk sistem UNIX.
    Archive of Reversing.ID
    Non-Cryptographic Hash
*/

// Fungsi ini merupakan method dari sebuah class
public long ELFHash(String key)
{
    long state = 0;
    long test  = 0;

    for(int i = 0; i < key.length(); i++)
    {
        state = (state << 4) + key.charAt(i);

        test = state & 0xF0000000L
        if(test != 0)
        {
            state ^= (x >> 24);
        }
        state &= ~x;
    }

    return state;
}